drgn_tools/rds.py

```diff
@@ -1020,6 +1117,421 @@ def rds_sock_info(
+                msg_list = cp.cp_retrans
+            else:
+                print("Invalid queue name provided in args !")
+                return None
+            # print(send_list)
+
+            for rm in list_for_each_entry(
+                "struct rds_message", msg_list.address_of_(), "m_conn_item"
+            ):
+                # print(rm)
+                lport_rm = be16_to_host(prog, rm.m_inc.i_hdr.h_sport)
+                rport_rm = be16_to_host(prog, rm.m_inc.i_hdr.h_dport)
+                seq = be64_to_host(prog, rm.m_inc.i_hdr.h_sequence)
+                num_bytes = be32_to_host(prog, rm.m_inc.i_hdr.h_len)
+
+                if laddr and conn_laddr not in laddr_list:
+                    continue
+                if raddr and conn_raddr not in raddr_list:
+                    continue
+                if tos and str(conn_tos) not in tos_list:
+                    continue
+                if lport and str(lport_rm) not in lport_list:
+                    continue
+                if rport and str(rport_rm) not in rport_list:
+                    continue
+
+                index = index + 1
+                msg_queue.append(
+                    [
+                        index,
+                        conn_laddr,
+                        lport_rm,
+                        conn_raddr,
+                        rport_rm,
+                        conn_tos,
+                        seq,
+                        num_bytes,
+                    ]
+                )
+
+    if queue.lower() == "send":
+        print("Send Message Queue:")
+    elif queue.lower() == "retrans":
+        print("Retransmit Message Queue:")
+
+    print_table(msg_queue, outfile, report)
+    print("\n")
+    return None
+
+
+def rds_print_msg_queue(
+    prog: drgn.Program,
+    queue: str = "All",
+    laddr: Optional[str] = None,
+    raddr: Optional[str] = None,
+    tos: Optional[str] = None,
+    lport: Optional[str] = None,
+    rport: Optional[str] = None,
+    ret: Optional[bool] = False,
+    outfile: Optional[str] = None,
+    report: bool = False,
+) -> None:
+    """
+    Print the rds msg queue similar rds-info -srt
+
+    :param prog: drgn program
+    :prarm queue: The msg queue to be displaed. Ex: 'send', 'recv' , 'retrans', 'All'. Default: 'All'
+    :param laddr: comma seperated string list of LOCAL-IP.  Ex: '192.168.X.X, 10.211.X.X, ...'
+    :param raddr: comma seperated string list of REMOTE-IP.  Ex: '192.168.X.X, 10.211.X.X, ...'
+    :param tos: comma seperated string list of TOS.  Ex: '0, 3, ...'
+    :param lport: comma seperated string list of lport.  Ex: 2259, 36554, ...'
+    :param rport: comma seperated string list of rport.  Ex: 2259, 36554, ...'
+    :param outfile: A file to write the output to.
+    :param report: Open the file in append mode. Used to generate a report of all the functions in the rds module.
+    :returns: None
+
+    """
+    msg = ensure_debuginfo(prog, ["rds"])
+    if msg:
+        print(msg)
```

Comment:
Do not use `ensure_debuginfo()` inside a corelens module code

Reply:
Hey Stephen, is there any specific reason for this ? 

- When authoring the code, we had discussed that it was best practice to ensure the rds module in the functions. Has this changed ?

Reply:
I have two minor issues with ensure_debuginfo():

1. It's not 100% accurate at telling whether a module has debuginfo loaded. That can't really be helped though, and in practice it works well enough.
2. Every time the function is called, we have to iterate over the list of kernel modules in the vmcore. Even if debuginfo is loaded for that module. This may not be cheap, especially on live systems. So the first priority with ensure_debuginfo() is to never call it in a loop, or in small helper functions that get called repeatedly.

Overall, it's not terrible but I worry about unnecessary performance issues. But that's not really your problem in this code, so I'll drop the issue. I think it would be good to cache the result of ensure_debuginfo() to make it less of an issue when calls are repeated.